latch_cli/services/cp/upload.py

```diff
@@ -299,70 +287,94 @@ class StartUploadReturnType:
-        src = src.resolve()
-
-    content_type, _ = mimetypes.guess_type(src)
-    if content_type is None:
-        with open(src, "rb") as f:
-            sample = f.read(units.KiB)
-
-        try:
-            sample.decode()
-            content_type = "text/plain"
-        except UnicodeDecodeError:
-            content_type = "application/octet-stream"
-
-    file_size = src.stat().st_size
-    if file_size > latch_constants.maximum_upload_size:
-        raise ValueError(
-            f"File is {with_si_suffix(file_size)} which exceeds the maximum upload size"
-            " (5TiB)"
+    try:
+        if not src.exists():
+            raise ValueError(f"Could not find {src}: no such file or link")
+
+        if src.is_symlink():
+            src = src.resolve()
+
+        content_type, _ = mimetypes.guess_type(src)
+        if content_type is None:
+            with open(src, "rb") as f:
+                sample = f.read(units.KiB)
+
+            try:
+                sample.decode()
+                content_type = "text/plain"
+            except UnicodeDecodeError:
+                content_type = "application/octet-stream"
+
+        file_size = src.stat().st_size
+        if file_size > latch_constants.maximum_upload_size:
+            raise ValueError(
+                f"File is {with_si_suffix(file_size)} which exceeds the maximum upload"
+                " size (5TiB)"
+            )
+
+        part_count = min(
+            latch_constants.maximum_upload_parts,
+            math.ceil(file_size / latch_constants.file_chunk_size),
+        )
+        part_size = max(
+            latch_constants.file_chunk_size,
+            math.ceil(file_size / latch_constants.maximum_upload_parts),
         )
 
-    part_count = min(
-        latch_constants.maximum_upload_parts,
-        math.ceil(file_size / latch_constants.file_chunk_size),
-    )
-    part_size = max(
-        latch_constants.file_chunk_size,
-        math.ceil(file_size / latch_constants.maximum_upload_parts),
-    )
+        retries = 0
+        while retries < MAX_RETRIES:
+            if throttle is not None:
+                time.sleep(throttle.get_delay())
 
-    res = tinyrequests.post(
-        latch_config.api.data.start_upload,
-        headers={"Authorization": get_auth_header()},
-        json={
-            "path": dest,
-            "content_type": content_type,
-            "part_count": part_count,
-        },
-    )
+            start = time.monotonic()
+            res = tinyrequests.post(
+                latch_config.api.data.start_upload,
+                headers={"Authorization": get_auth_header()},
+                json={
+                    "path": dest,
+                    "content_type": content_type,
+                    "part_count": part_count,
+                },
+            )
+            end = time.monotonic()
 
-    json_data = res.json()
+            if latency_q is not None:
+                latency_q.put(end - start)
```

Comment:
Idk why this is optional but idrc

Reply:
in the case where u upload a single file, none of the concurrency stuff is passed to the function